### PR TITLE
#34923: Desktop site bootstrap

### PR DESCRIPTION
**Pull request by jfboismenu**
In the process of writing unit tests for the boostrap code in Toolkit, I found a series of bugs in Mockgun which I have fixed. I've also provided unit tests.

### CODE REVIEW

File: tests/test_mockgun.py [hunk 1]
```diff
@@ -42,19 +42,15 @@
...
 
-def setUpModule():
-    """
-    Sets the Shotgun schema for this series of unit tests.
-    """
-    mockgun_schema_folder = os.path.join(
-        os.path.dirname(__file__),
-        "mockgun"
-    )
+mockgun_schema_folder = os.path.join(
```
Comment (jfboismenu) [inline tests/test_mockgun.py]: Python 2.5's unit tests module doesn't support the setUpModule method, so do it globally instead.

File: shotgun_api3/lib/mockgun.py [hunk 2]
```diff
@@ -560,6 +560,22 @@ def _get_new_row(self, entity_type):
...
+        :param lval: Value inside that field. Can be of any type: datetime, date, int, str, bool, etc.
+        :param str operator: Name of the operator to use.
+        :param rval: The value following the operator in a filter.
+
+        :returns: The result of the operator that was applied.
+        :rtype: bool
+        """
+        # If we have a list of scalar values
+        if isinstance(lval, list) and field_type != "multi_entity":
+            # Compare each one. If one matches the predicatet we're good!
```
Comment (josh-t) [inline shotgun_api3/lib/mockgun.py]: typo: predicate

File: shotgun_api3/lib/mockgun.py [hunk 3]
```diff
@@ -609,8 +625,23 @@ def _compare(self, field_type, lval, operator, rval):
                 return lval.endswith(rval)
         elif field_type == "entity":
             if operator == "is":
+                # If the field is set to None
+                if lval is None:
+                    # Return true if rval is None too.
+                    return rval is None
+                if rval is None:
+                    # We already know lval is not None, so we know they are not equal.
+                    return False
```
Comment (josh-t) [inline shotgun_api3/lib/mockgun.py]: Could simplify:
```
if None in [rval, lval]:
    return rval is lval
```

File: shotgun_api3/lib/mockgun.py [hunk 4]
```diff
@@ -609,8 +625,23 @@ def _compare(self, field_type, lval, operator, rval):
...
+                # Both values are set, compare them.
                 return lval["type"] == rval["type"] and lval["id"] == rval["id"]
             elif operator == "is_not":
+                # If the field is set to None
+                if lval is None:
+                    # Return true if rval is not None.
+                    return rval is not None
+                if rval is None:
+                    # We already know lval is not None, so we know they are not equal.
+                    return True
```
Comment (josh-t) [inline shotgun_api3/lib/mockgun.py]: Similarly:
```
if None in [rval, lval]:
    return rval is not lval
```

File: shotgun_api3/lib/mockgun.py [hunk 5]
```diff
@@ -695,17 +740,21 @@ def _row_matches_filter(self, entity_type, row, sg_filter, retired_only):
...
             # if we're operating on an entity, we'll need to grab the name from the lval's row
             if field_type == "entity":
-                lval_row = self._db[lval["type"]][lval["id"]]
-                if "name" in lval_row:
-                    lval["name"] = lval_row["name"]
-                elif "code" in lval_row:
-                    lval["name"] = lval_row["code"]
+                # If the entity field is set, we'll retrieve the name of the entity.
+                if lval is not None:
+                    lval_row = self._db[lval["type"]][lval["id"]]
```
Comment (josh-t) [inline shotgun_api3/lib/mockgun.py]: nitpick. difficult to parse visually. maybe break out the key string into a separate variable?

**Pull request closed without merging**